services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java

```diff
@@ -371,13 +371,12 @@ protected Response handleLoginResponse(String samlResponse, SAMLDocumentHolder h
                     assertionElement = DocumentUtil.getElement(holder.getSamlDocument(), new QName(JBossSAMLConstants.ASSERTION.get()));
                 }
 
-                if (config.isWantAssertionsSigned() && config.isValidateSignature()) {
-                    if (!AssertionUtil.isSignatureValid(assertionElement, getIDPKeyLocator())) {
-                        logger.error("validation failed");
-                        event.event(EventType.IDENTITY_PROVIDER_RESPONSE);
-                        event.error(Errors.INVALID_SIGNATURE);
-                        return ErrorPage.error(session, Messages.INVALID_REQUESTER);
-                    }
+                if ((config.isWantAssertionsSigned() && !AssertionUtil.isSignedElement(assertionElement))
+                        || (config.isValidateSignature() && !AssertionUtil.isSignatureValid(assertionElement, getIDPKeyLocator()))) {
+                    logger.error("validation failed");
```

Comment:
I don't understand your question. The changed code checks for validity of Assertion element regardless of the wantAssertionsSigned flag. The IdP can be configured to sign the whole document only, not individual assertions. Is this case still working? The PR has to contain an integration test for that.